#161340178 Build out rest endpoints for articles (CRUD)

nakatuddesuzan opened the pull request:
#### What does this PR do?
Users should be able to create articles
#### Description of Task to be completed?
- A user should be able to post an article
- Also, ensure that the following endpoints are working
1. ##### GET api/articles/
2. ##### GET api/articles/<slug>
3. ##### DELETE api/articles/<slug>
4. ##### PUT api/articles/<slug>
#### How should this be manually tested?
- Install the necessary dependancies
- Test the endpoints using postman by using those urls
- Use postman to test each endpoint with the headers
     - key:AUTHORIZATION value: bearer <token>
     - key:Content-Type value: application/json
#### What are the relevant pivotal tracker stories?
[#161340178](https://www.pivotaltracker.com/story/show/161340178)

#### Screenshots
- How to add an article
<img width="461" alt="screen shot 2018-11-08 at 20 09 42" src="https://user-images.githubusercontent.com/39129473/48214991-4eb4c400-e392-11e8-9269-e547f64df4d2.png">

- How data is presented when all articles are got
<img width="1067" alt="screen shot 2018-11-08 at 20 10 38" src="https://user-images.githubusercontent.com/39129473/48215050-68eea200-e392-11e8-89e8-359d89aa8542.png">

Musacoli commented on authors/apps/articles/migrations/0001_initial.py:
```diff
@@ -0,0 +1,35 @@
+import taggit.managers
+
+
+class Migration(migrations.Migration):
```
I think it would be best not to push the migrations to GitHub and ignore the entire migrations file as earlier agreed.

Please look into this

nakatuddesuzan replied:
I had issues in on my local machine, so I had to delete the database, so to run migrations again I had to delete the migrations folder and run them when I pushed to GitHub, Travis started failing complaining about not seeing some migrations file, which made me push the folder. This won't happen again.

Musacoli commented on authors/apps/articles/models.py:
```diff
@@ -0,0 +1,23 @@
+    published = models.BooleanField()
+    updated_at = models.DateTimeField(auto_now=True)
+    author = models.ForeignKey(User, on_delete=models.CASCADE)
+    tags = TaggableManager(blank=True)
```
In regards to this section, I think it would be best if we aligned our code with what the documentation entails so that it's easier to have other team members use it with ease .

So my suggestion would be to change this to 

`"tagList":`

krmroland commented on authors/apps/articles/serializers.py:
```diff
@@ -0,0 +1,71 @@
+        fields = ['title','description', 'body', 'author', 
+                    'created_at', 'updated_at', 'tags', 'slug', 'published', 'image']
+
+    def validate_title(self, value):
```
The title definition in the model already has the `max_length` attribute, you could as well as add the `min-length` if you wanted to validate that constraint so that you don't have to write the entire validation logic.

nakatuddesuzan replied:
I really don't think we need to limit people here,  let me give an example when you are to post on medium.com, even if its one letter you cannot be limited to have that, but for the sake of maximum length, the title doesn't really have to be so long. It can't be like a body

Musacoli commented on authors/apps/articles/serializers.py:
```diff
@@ -0,0 +1,71 @@
+        model = Article
+
+        fields = ['title','description', 'body', 'author', 
+                    'created_at', 'updated_at', 'tags', 'slug', 'published', 'image']
```
I would also advise that you implement the tagList change here to as it affects the article posting documentation we are currently using

`{
  "article": {
    "title": "How to train your dragon",
    "description": "Ever wonder how?",
    "body": "You have to believe",
    "tagList": ["reactjs", "angularjs", "dragons"]
  }
}`

krmroland commented on authors/apps/articles/serializers.py:
```diff
@@ -0,0 +1,71 @@
+        return value
+
+class UpdateArticleAPIVIEWSerializer(serializers.ModelSerializer):
+    # user_id = User.pk
```
Is there any chance we would need this commented line in future?  if no then probably it shouldn't have been committed

nakatuddesuzan replied:
Fixed

Musacoli reviewed (changes requested):
My other suggestion would be that you also find a way to set the `published` column to automatically populate itself with true once a user has published an article since at the moment it has to be done manually.

But otherwise great work done quite impressive how you got so much done in such little time.

krmroland commented on authors/apps/articles/serializers.py:
```diff
@@ -0,0 +1,71 @@
+            )
+        return value
+    def validate_description(self, value):
+        if len(value) > 200:
```
Same for description, Could there be a specific reason why we are not relying on the native validation functionality available in the `Django Rest framework`?

krmroland commented on authors/apps/articles/views.py:
```diff
@@ -0,0 +1,67 @@
+        serializer.is_valid(raise_exception=True)
+        serializer.save(author=user_info[0])
+        data = serializer.data
+        data["message"] = "Your Article has been created successfully"
```
This is a no big deal, but massaging of the data to be returned could be done in the `Renderer classes` since you already have one.

nakatuddesuzan replied:
Okay

krmroland commented on authors/settings.py:
```diff
@@ -43,12 +43,15 @@
     'django.middleware.security.SecurityMiddleware',
     'django.middleware.security.SecurityMiddleware',
-    'whitenoise.middleware.WhiteNoiseMiddleware',
+    # 'whitenoise.middleware.WhiteNoiseMiddleware',
```
Could there be any specific reason why you commented out `whitenoise.middleware.WhiteNoiseMiddleware`

nakatuddesuzan replied:
Fixed

krmroland commented on requirements.txt:
```diff
@@ -1,21 +1,51 @@
+django-taggit-serializer==0.1.7
 djangorestframework==3.9.0
+docopt==0.6.2
+flake8==3.6.0
```
The number of packages in the `requirements.txt` has massively grown. Could they be dependencies of some package(s) you installed ? or is there a possibility that you forgot and did a `pip freeze` in a global environment?

nakatuddesuzan replied:
I installed them, I did not pip freeze in a global environment, they are dependencies

krmroland reviewed (changes requested):
Hey @nakatuddesuzan . I was really impressed by how you implemented the Tags Feature.

I have shared some few opinions on where I think we could do better.

SoniaRMK commented:
@nakatuddesuzan kindly work on the feedback given by your colleagues as well as the PR description per our discussion yesterday.

krmroland commented:
Hey @nakatuddesuzan , I think you forgot to squash your commits to one.

Commit:
feat(create articles):Users should be able to create articles

- fixed bug in the update
- write tests for the functionality
- add base test file

[#161340178]
